internal/task/cgroups/cgroup.go

```diff
@@ -0,0 +1,102 @@
+package cgroups
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+	"time"
+)
+
+const baseCgroupPath = "/sys/fs/cgroup/"
+
+// CreateCgroupForTask creates a cgroup for a task
+// In a real system, we would want the cgroup config to be configurable
+func CreateCgroupForTask(taskID string) (*os.File, error) {
+	cgroupPath := filepath.Join(baseCgroupPath, taskID)
+
+	if err := os.MkdirAll(cgroupPath, 0755); err != nil {
+		return nil, fmt.Errorf("failed to create cgroup directory %s: %w", cgroupPath, err)
+	}
+
+	// Configure CPU limits: Quota 200000 µs and Period 1000000 µs.
+	// this should cap the cpu usage to 20%
+	cpuMaxPath := filepath.Join(cgroupPath, "cpu.max")
+	cpuConfig := "200000 1000000"
+	if err := os.WriteFile(cpuMaxPath, []byte(cpuConfig), 0644); err != nil {
+		return nil, fmt.Errorf("failed to write CPU config to %s: %w", cpuMaxPath, err)
+	}
+
+	// Configure memory limit: 64M.
+	memoryMaxPath := filepath.Join(cgroupPath, "memory.max")
+	memoryConfig := "64M"
+	if err := os.WriteFile(memoryMaxPath, []byte(memoryConfig), 0644); err != nil {
+		return nil, fmt.Errorf("failed to write memory config to %s: %w", memoryMaxPath, err)
+	}
+
+	// io is not always enabled on the system and can be enabled by:
+	// echo "+io" | sudo tee /sys/fs/cgroup/cgroup.subtree_control
+	// Configure IO limits: device "8:0" with max read and write bandwidth 1048576 (1 MB/s).
+	ioMaxPath := filepath.Join(cgroupPath, "io.max")
+	ioConfig := "8:0 rbps=1048576 wbps=1048576"
+	if err := os.WriteFile(ioMaxPath, []byte(ioConfig), 0644); err != nil {
+		return nil, fmt.Errorf("failed to write IO config to %s: %w", ioMaxPath, err)
+	}
```

Comment:
This applies to most of these cgroup controller writes, but I'd suggest adding a method to clean up hanging cgroups if one of these writes fails. For example, the `io` controller wasn't enabled on my box initially, so this write failed. The resulting cgroups don't get cleaned up in that case, and pile up pretty rapidly.